util/src/main/java/org/killbill/billing/util/optimizer/BusDispatcherOptimizerOn.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020-2021 Equinix, Inc
+ * Copyright 2014-2021 The Billing Project, LLC
+ *
+ * The Billing Project licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.billing.util.optimizer;
+
+import org.killbill.billing.callcontext.InternalCallContext;
+import org.killbill.billing.events.BusInternalEvent;
+import org.killbill.billing.util.callcontext.CallOrigin;
+import org.killbill.billing.util.callcontext.InternalCallContextFactory;
+import org.killbill.billing.util.callcontext.UserType;
+import org.killbill.billing.util.config.definition.EventConfig;
+import org.killbill.bus.api.BusEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+
+public class BusDispatcherOptimizerOn implements BusDispatcherOptimizer{
+
+    private static final Logger logger = LoggerFactory.getLogger(BusDispatcherOptimizerOn.class);
+
+    private final EventConfig eventConfig;
+    private final InternalCallContextFactory internalCallContextFactory;
+
+    @Inject
+    public BusDispatcherOptimizerOn(EventConfig eventConfig, final InternalCallContextFactory internalCallContextFactory) {
+        this.eventConfig = eventConfig;
+        this.internalCallContextFactory = internalCallContextFactory;
+    }
+
+    @Override
+    public boolean shouldDispatch(final BusEvent event) {
+        Preconditions.checkState(event instanceof BusInternalEvent, "Unexpected external bus event %s, skip...", event);
```

Comment:
Technically, this won't _skip_: the exception in the handler will trigger a re-dispatch. Wondering if logging and returning wouldn't be better here?

Reply:
It should not happen (otherwise it's a bug and we have to fix it). The precondition will just make it obvious. Agree that `skip` is a bit misleading, i could change it to something like `corruption`, ... ?

Reply:
FWIW I am more worried about throwing the exception (and re-dispatching the event 3 times) vs the log message.

Reply:
I personally don't care either way, if this happens, it's a bug in the wiring (only internal bus should use the `BusOptimizer `) hence the comment [here](https://github.com/sbrossie/killbill/blob/events-optimization/beatrix/src/main/java/org/killbill/billing/beatrix/extbus/BeatrixListener.java#L89).

I think all we care is that it's visible (in a way 3 retries will be even more visible), so we can fix it. If you have a suggestion on specific log you want to see, write it down, and I 'll update the code.